### PR TITLE
PIM-5589: implement channel import with new system (prepare legacy remove)

### PR DESCRIPTION
**Pull request by nidup**
To be able to totally remove the deprecated import system (pre v1.4), we need to implement missing imports used by installer and behat context.
Here is a PR about implementing the channel import.
You'll find more details in the related Jira issue.
Should be merged in future 1.6 (dev-master).

TODO:
 - [x] fix EE
 - [x] rebase

| Q                 | A
| ----------------- | ---
| Specs             | Y
| Behats            | N
| Changelog updated | Y
| Review and 2 GTM  | Y

### CODE REVIEW

File: src/Pim/Bundle/InstallerBundle/Resources/config/batch_jobs.yml [hunk 1]
```diff
@@ -36,21 +36,21 @@ connector:
...
                         charsetValidator: pim_connector.validator.item.charset_validator
                 import:
-                    title:     pim_installer.jobs.fixtures_channel_yml.import.title
+                    title:     pim_installer.jobs.fixtures_channel_csv.import.title
                     services:
-                        reader:    pim_base_connector.reader.file.yaml
-                        processor: pim_base_connector.processor.channel
-                        writer:    pim_base_connector.writer.doctrine
+                        reader:    pim_connector.reader.file.csv_channel
+                        processor: pim_connector.processor.denormalization.channel.flat
```
Comment (juliensnz) [inline src/Pim/Bundle/InstallerBundle/Resources/config/batch_jobs.yml]: why flat ? it should be the pivot format no ? Do you use a csv converter ?

File: src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        $this->updater        = $updater;
+        $this->validator      = $validator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process($item)
+    {
+        $convertedItem = $this->convertItemData($item);
```
Comment (juliensnz) [inline src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php]: as previously discussed we should put this part in the reader

File: src/Pim/Bundle/InstallerBundle/Resources/config/batch_jobs.yml (hunk 1 above)
Comment (nidup) [inline src/Pim/Bundle/InstallerBundle/Resources/config/batch_jobs.yml]: as any other processor, yes it is a standard processor but converting flat/csv to standard format via a converter. anyway this converter could be into the reader, in this case we could remove the flat notation here.

File: src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php (hunk 2 above)
Comment (nidup) [inline src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php]: 1000% agree, I do this for consistency to prepare the the drop of the deprecated system but i would love move the array converter in the reader for any import as we imagined at the very beginning :)

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    /**
+     * @param ChannelInterface $channel
+     * @param string           $field
+     * @param mixed            $data
+     *
+     * @throws \InvalidArgumentException
+     */
+    protected function setData(ChannelInterface $channel, $field, $data)
+    {
+        if ('code' === $field) {
```
Comment (solivier) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: maybe use a switch case ?

**Comment by solivier** [comment]
gtm after comment fixed

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            $channel->setCode($data);
+        }
+
+        if ('tree' === $field) {
+            $category = $this->findCategory($data);
+            if (null !== $category) {
+                $channel->setCategory($category);
+            } else {
+                throw new \InvalidArgumentException(sprintf('Category with "%s" code does not exist', $data));
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: You should not throw this here. This exception is raised when:
> Exception thrown if an argument is not of the *expected type*.

It's time to introduce a business dedicated UpdateException ;)

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            }
+        }
+
+        if ('locales' === $field) {
+            foreach ($data as $localeCode) {
+                $locale = $this->findLocale($localeCode);
+                if (null !== $locale) {
+                    $channel->addLocale($locale);
+                } else {
+                    throw new \InvalidArgumentException(sprintf('Locale with "%s" code does not exist', $localeCode));
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: same here

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            }
+        }
+
+        if ('currencies' === $field) {
+            foreach ($data as $currencyCode) {
+                $currency = $this->findCurrency($currencyCode);
+                if (null !== $currency) {
+                    $channel->addCurrency($currency);
+                } else {
+                    throw new \InvalidArgumentException(sprintf('Currency with "%s" code does not exist', $currencyCode));
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: same here

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+            $channel->setColor($data);
+        }
+    }
+
+    /**
+     * @param string $code
+     *
+     * @return CategoryInterface|null
+     */
+    protected function findCategory($code)
```
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: imvho all these one line methods are useless and make the code more complicated

File: spec/Pim/Component/Catalog/Updater/ChannelUpdaterSpec.php [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        ChannelInterface $channel,
+        CategoryInterface $tree,
+        LocaleInterface $enUS,
+        LocaleInterface $frFR,
+        CurrencyInterface $usd,
+        CurrencyInterface $eur
+    ) {
+        $values = [
+            'code'  => 'ecommerce',
+            'label' => 'Ecommerce',
```
Comment (pierallard) [inline spec/Pim/Component/Catalog/Updater/ChannelUpdaterSpec.php]: Indentation

File: spec/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverterSpec.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+            'color'        => 'orange'
+        ];
+
+        $result = [
+            'code'         => 'ecommerce',
+            'label'        => 'Ecommerce',
+            'locales'      => ['en_US', 'fr_FR'],
+            'currencies'   => ['EUR', 'USD'],
+            'tree'         => 'master_catalog',
+            'color'        => 'orange'
```
Comment (pierallard) [inline spec/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverterSpec.php]: too many indentation

File: src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+            $convertedItem[$field] = explode(',', $data);
+        }
+
+        return $convertedItem;
+    }
+
+    /**
+     * @param array $item
+     */
+    protected function validate(array $item)
```
Comment (jjanvier) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php]: useless method imvho

File: spec/Pim/Component/Connector/Processor/Denormalization/ChannelProcessorSpec.php [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+               'code'         => 'mycode',
+               'label'        => 'Ecommerce',
+               'locales'      => 'en_US,fr_FR',
+               'currencies'   => 'EUR,USD',
+               'tree'         => 'master_catalog',
+               'color'        => 'orange'
+            ],
+            'converted_values' => [
+              'code'   => 'mycode',
+              'label'  => 'Ecommerce',
```
Comment (pierallard) [inline spec/Pim/Component/Connector/Processor/Denormalization/ChannelProcessorSpec.php]: identation

File: src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+        return $channel;
+    }
+
+    /**
+     * @param array $item
+     *
+     * @return array
+     */
+    protected function convertItemData(array $item)
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php]: same comment as before, those one line methods serve no purpose imho

File: src/Pim/Bundle/InstallerBundle/Resources/fixtures/icecat_demo_dev/channels.csv [hunk 13]
```diff
@@ -0,0 +1,4 @@
+code;label;locales;currencies;tree;color
+mobile;Mobile;fr_FR,de_DE,en_US;USD,EUR;master;blue
+print;Print;fr_FR,de_DE,en_US;USD,EUR;master;gray
+ecommerce;Ecommerce;fr_FR,de_DE,en_US;USD,EUR;master;orange
```
Comment (pierallard) [inline src/Pim/Bundle/InstallerBundle/Resources/fixtures/icecat_demo_dev/channels.csv]: No newline ? (don't know if it's mandatory in CSV files)

File: src/Pim/Component/Catalog/Factory/ChannelFactory.php [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * @param string $channelClass
+     */
+    public function __construct($channelClass)
+    {
+        $this->channelClass = $channelClass;
+    }
+
+    /**
+     * @return ChannelInterface
```
Comment (pierallard) [inline src/Pim/Component/Catalog/Factory/ChannelFactory.php]: Doc ?

File: src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+                        $requiredField,
+                        implode(', ', array_keys($item))
+                    )
+                );
+            }
+
+            if ('' === $item[$requiredField]) {
+                throw new ArrayConversionException(
+                    sprintf(
+                        'Field "%s" must be filled',
```
Comment (skeleton) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php]: 1 `%s` is missing (there are 2 arguments in your sprintf())

File: src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+    protected $updater;
+
+    /** @var ValidatorInterface */
+    protected $validator;
+
+    /** @var ChannelFactory */
+    protected $channelFactory;
+
+    /**
+     * @param StandardArrayConverterInterface       $channelConverter
```
Comment (skeleton) [inline src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php]: wrong phpdoc

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+            }
+        }
+
+        if ('label' === $field) {
+            $channel->setLabel($data);
+        }
+
+        if ('color' === $field) {
+            $channel->setColor($data);
+        }
```
Comment (pierallard) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: Agree with @solivier 's comment. You check n times 'xx' == $field, you should use a case-when.

File: src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * @param array  $convertedItem
+     * @param string $field
+     * @param mixed  $data
+     *
+     * @return array
+     */
+    protected function convertField(array $convertedItem, $field, $data)
+    {
+        if ('code' === $field || 'color' === $field || 'tree' === $field || 'label' === $field) {
```
Comment (pierallard) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php]: in_array($field, ['code'...])

File: src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php (hunk 12 above)
Comment (jjanvier) [inline src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php]: plus it adds extra code for nothing, which makes it more complicated

File: src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+
+        try {
+            $this->updateChannel($channel, $convertedItem);
+        } catch (\InvalidArgumentException $exception) {
+            $this->skipItemWithMessage($item, $exception->getMessage(), $exception);
+        }
+
+        $violations = $this->validateChannel($channel);
+        if ($violations->count() > 0) {
+            $this->skipItemWithConstraintViolations($item, $violations);
```
Comment (skeleton) [inline src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php]: can we detach channel if it's invalid ? (maybe it's useless, dunno...)

File: src/Pim/Component/Catalog/Factory/ChannelFactory.php [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+ * Class ChannelFactory
+ *
+ * @author    Nicolas Dupont <nicolas@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ChannelFactory
+{
+    /** @var ChannelInterface */
+    protected $channelClass;
```
Comment (willy-ahva) [inline src/Pim/Component/Catalog/Factory/ChannelFactory.php]: should be "@var string class must implement ChannelInterface"

File: src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+                    )
+                );
+            }
+
+            if ('' === $item[$requiredField]) {
+                throw new ArrayConversionException(
+                    sprintf(
+                        'Field "%s" must be filled',
+                        $requiredField,
+                        implode(', ', array_keys($item))
```
Comment (pierallard) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php]: You missed an argument in your string

File: src/Pim/Component/Catalog/Factory/ChannelFactory.php (hunk 14 above)
Comment (willy-ahva) [inline src/Pim/Component/Catalog/Factory/ChannelFactory.php]: seems ok

**Comment by pierallard** [comment]
GTM after comment fix

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+     * @param IdentifiableObjectRepositoryInterface $localeRepository
+     * @param IdentifiableObjectRepositoryInterface $currencyRepository
+     */
+    public function __construct(
+        IdentifiableObjectRepositoryInterface $categoryRepository,
+        IdentifiableObjectRepositoryInterface $localeRepository,
+        IdentifiableObjectRepositoryInterface $currencyRepository
+    ) {
+        $this->categoryRepository = $categoryRepository;
+        $this->localeRepository = $localeRepository;
```
Comment (willy-ahva) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: align

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 3 above)
Comment (willy-ahva) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: :+1:

File: src/Pim/Bundle/InstallerBundle/Resources/fixtures/icecat_demo_dev/channels.csv (hunk 13 above)
Comment (solivier) [inline src/Pim/Bundle/InstallerBundle/Resources/fixtures/icecat_demo_dev/channels.csv]: @pierallard there is a new line

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 4 above)
Comment (willy-ahva) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: or `\UnexpectedValueException` ?

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 17 above)
Comment (jjanvier) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: (I prefer elseif personally, but no big deal ^^)

File: src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+     * @param mixed  $data
+     *
+     * @return array
+     */
+    protected function convertField(array $convertedItem, $field, $data)
+    {
+        if ('code' === $field || 'color' === $field || 'tree' === $field || 'label' === $field) {
+            $convertedItem[$field] = $data;
+        } else {
+            $convertedItem[$field] = explode(',', $data);
```
Comment (willy-ahva) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php]: If you have data that are `en_US, fr_FR` you will have extra spaces in your values. Add trim on values returned by explode.

File: src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+        IdentifiableObjectRepositoryInterface $repository,
+        StandardArrayConverterInterface $channelConverter,
+        ChannelFactory $channelFactory,
+        ObjectUpdaterInterface $updater,
+        ValidatorInterface $validator
+    ) {
+        parent::__construct($repository);
+
+        $this->arrayConverter = $channelConverter;
+        $this->channelFactory  = $channelFactory;
```
Comment (willy-ahva) [inline src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php]: extra space before =

**Comment by willy-ahva** [comment]
GTM after comments fixed

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 4 above)
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: @jjanvier, yes good time for business exception

File: src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php (hunk 12 above)
Comment (nidup) [inline src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php]: i copied from other processor, i guess original idea was to ease overriding a part which is ... at the end ... not even a good idea :)

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 17 above)
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: i prefer elseif too (i change), switch case with 'break' stuff looks imvho less readable

File: src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php (hunk 19 above)
Comment (nidup) [inline src/Pim/Component/Connector/Processor/Denormalization/ChannelProcessor.php]: yes we should, good catch (even if there is very few chances to create memleak by importing channels except if you have bunch of them ;) )

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 3 above)
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: i do prefer if elseif (more readable imho)

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 22 above)
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: style comment + not in cs fixer = not a rule, i won't fix :)

File: src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php (hunk 15 above)
Comment (nidup) [inline src/Pim/Component/Connector/ArrayConverter/Flat/ChannelStandardConverter.php]: fixed, thx

File: src/Pim/Component/Catalog/Factory/ChannelFactory.php (hunk 14 above)
Comment (nidup) [inline src/Pim/Component/Catalog/Factory/ChannelFactory.php]: is ok, i don't add text as it's quite obvious that it created a channel :)

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 6 above)
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: k

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 5 above)
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: k

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 4 above)
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: ok i will do a second PR doing this for any updaters

File: src/Pim/Component/Catalog/Updater/ChannelUpdater.php (hunk 7 above)
Comment (nidup) [inline src/Pim/Component/Catalog/Updater/ChannelUpdater.php]: using identifiableobjectrepositoryinterface it allows to typehint returned object which is maybe not a good reason

File: spec/Pim/Component/Connector/Processor/Denormalization/ChannelProcessorSpec.php (hunk 11 above)
Comment (nidup) [inline spec/Pim/Component/Connector/Processor/Denormalization/ChannelProcessorSpec.php]: fixed

File: spec/Pim/Component/Catalog/Updater/ChannelUpdaterSpec.php (hunk 8 above)
Comment (nidup) [inline spec/Pim/Component/Catalog/Updater/ChannelUpdaterSpec.php]: fixed

File: src/Pim/Bundle/InstallerBundle/Resources/fixtures/icecat_demo_dev/channels.csv (hunk 13 above)
Comment (nidup) [inline src/Pim/Bundle/InstallerBundle/Resources/fixtures/icecat_demo_dev/channels.csv]: yep there is

Comment (pierallard) [inline src/Pim/Bundle/InstallerBundle/Resources/fixtures/icecat_demo_dev/channels.csv]: Github is displaying a "no newline at end of file"...

**Pull request merged**